// @geonode/repocket-peer-client/dist/src/services/api.service.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiService = void 0;
const env_1 = require("../shared/env");
const http_service_1 = __importDefault(require("./http.service"));
class ApiService {
    constructor(path = '', options = { baseUrl: env_1.ENV_CONFIG.apiUrl }) {
        this.baseApi = `${options.baseUrl}/${path}`;
        this.httpService = new http_service_1.default();
    }
    get(endpoint = '', params) {
        return this.httpService.get(`${this.baseApi}${endpoint}`, params);
    }
    post(endpoint = '', payload) {
        return this.httpService.post(`${this.baseApi}${endpoint}`, payload);
    }
    put(endpoint = '', payload) {
        return this.httpService.put(`${this.baseApi}${endpoint}`, payload);
    }
    delete(endpoint = '', data) {
        return this.httpService.delete(`${this.baseApi}${endpoint}`, data);
    }
}
exports.ApiService = ApiService;
